doc/guides/4 - Refinery Extensions/7 - changes to extensions in Refinery 3.0.textile

```diff
@@ -0,0 +1,194 @@
+<code>
+// refinerycms-blog/app/assets/stylesheets/blog/refinery/backend.css.scss
+
+@import 'refinery/mixins';
+@import 'refinery/colours';
+@import 'font-awesome-sprockets', 'font-awesome';
+@import 'refinery/icons';
+
+.comments_icon {
+  @include icon('comments')
+}
+...
+</code>
+</pre>
+
+
+
+h2. Assets
+
+To add your assets to the assets defined with Refinery use @config/initializers/assets.rb@. A simple assets.rb would look like this:
+
+<pre>
+<code>
+  Refinery::Core.configure do |config|
+    # Register extra javascripts for backend
+    config.register_javascript "refinerycms-testimonials.js"
+  
+    # Register extra stylesheets for backend 
+    config.register_stylesheet "refinerycms-testimonials", :media => 'screen'
+  end
+  Rails.application.config.assets.precompile += %w( refinerycms-testimonials.css refinerycms-testimonials.js )
+</code>
+</pre>
+
+Using "namespaced" filenames, as above, keeps your code linked to your extension and also helps to keep your stylesheets and javascripts from being overridden accidentally by users. 
+
+h2. Decorators and Strong Parameters
+
+Rails 4 requires the use of strong parameters. (You can turn them off altogether but that is not recommended)
+
+If your engine adds fields to a page, blog post or other defined object you need to add your fields to the page's list of acceptable parameters. Here is how to do that.
+
+<pre>
+<code>
+
+Refinery::Admin::PagesController.class_eval do
+
+  # work around an issue with stack level too deep, due to an issue with decorators.
+  if self.instance_methods.exclude?(:page_params_with_testimonial_params)
+    def page_params_with_testimonial_params
+
+      testimonial_params = params.require(:page).permit(:testimonials_show, :testimonials_count, :testimonials_select)
+      page_params_without_testimonial_params.merge(testimonial_params)
+
+    end
+
+    # Swap out the default page_params method with our new one
+    alias_method_chain :page_params, :testimonial_params
+  end
+end
+
+</code>
+</pre>
+
+
+
+
+h3. Page Decorators
+
+If you extension is producing an object or collection to be displayed on a page you can use decorators to add the object or collection to the page.
+
+Using @refinerycms-testimonials@ as an example again:
+
+<pre>
+<code>
+
+  # app/decorators/models/refinery/testimonials_page_decorator.rb
+
+  # Open the Refinery::Page model for manipulation 
+  Refinery::Page.class_eval do
+    attr_accessor :testimonials
+    validates :testimonials_count, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+  
+    def testimonials
+      if testimonials_show
+        n = testimonials_count==0 ? Refinery::Testimonials::Testimonial.count : testimonials_count
+        @testimonials = Refinery::Testimonials::Testimonial.all
+        @testimonials = testimonials_select=='Random' ?
+          @testimonials.order_by_rand.limit(n) :
+          @testimonials.order('created_at DESC').limit(n)
+      end
+    end
+  end
+</code>
+</pre>
```

Comment:
Yes...was planning to.
On Thu, 14 May 2015 at 5:26 am Philip Arndt <notifications@github.com>
wrote:

> In doc/guides/4 - Refinery Extensions/7 - changes to extensions in
> Refinery 3.0.textile
> <https://github.com/refinery/refinerycms/pull/2982#discussion_r30278166>:
>
> > +  Refinery::Page.class_eval do
> > +    attr_accessor :testimonials
> > +    validates :testimonials_count, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
> > +
> > +    def testimonials
> > +      if testimonials_show
> > +        n = testimonials_count==0 ? Refinery::Testimonials::Testimonial.count : testimonials_count
> > +        @testimonials = Refinery::Testimonials::Testimonial.all
> > +        @testimonials = testimonials_select=='Random' ?
> > +          @testimonials.order_by_rand.limit(n) :
> > +          @testimonials.order('created_at DESC').limit(n)
> > +      end
> > +    end
> > +  end
> > +</code>
> > +</pre>
>
> Can we use a simpler example please? [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/refinery/refinerycms/pull/2982/files#r30278166>.
>